src/SeaportLeverage.sol

```diff
@@ -0,0 +1,267 @@
+     * - The `orderType` must be `FULL_RESTRICTED`. This means only the `zone`,
+     * or the offerer, can fulfill the order.
+     * - The `conduitKey` must be zero. No conduit should be used.
+     * - The `totalOriginalConsiderationItems` must be 2.
+     *
+     * - The `Offer` item must be of type `ERC20`.
+     * - For the case of leverage, `token` of the `Offer` item must be the
+     * `COLLATERAL` token.
+     * - The `startAmount` and `endAmount` of the `Offer` item must be equal to
+     * `collateralToPurchase`. Start and end should be equal because the amount is fixed.
+     *
+     * - The first `Consideration` item must be of type `ERC20`.
+     * - The `token` of the first `Consideration` item must be this contract's
+     * address. This is to allow this contract to gain control flow. We also
+     * want to use the `transferFrom()` args to communicate data to the
+     * `transferFrom()` callback. Any data that can't be fit into the
+     * `transferFrom()` args will be communicated through transient storage.
+     * - The `startAmount` and `endAmount` of the first `Consideration` item
+     * communicate the amount of `BASE` asset to borrow from the IonPool during
+     * the callback.
+     * - The `recipient` of the first `Consideration` item must be `msg.sender`.
+     * This will be user's vault that will be deleveraged. This contract assumes
+     * that caller is the owner of the vault.
+     *
+     * - The second `Consideration` item must be of type `ERC20`.
+     * - The `token` of the second `Consideration` item must be the `BASE`
+     * - The second `Consideration` item must have the `startAmount` and `endAmount`
+     * equal to `amountToBorrow`.
+     *
+     * We don't constrain the `recipient` of the second `Consideration` item.
+     *
+     * It is technically possible for two distinct orders to have the same
+     * parameters. The `salt` should be used to distinguish between two orders
+     * with the same parameters. Otherwise, they will map to the same order hash
+     * and only one of them will be able to be fulfilled.
+     *
+     * @param order Seaport order.
+     * @param initialDeposit Amount of collateral to be transferred from sender. [WAD]
+     * @param resultingAdditionalCollateral Total collateral to be deposited from this call. [WAD]
+     * @param amountToBorrow Amount of base asset to borrow. [WAD]
+     * @param proof Merkle path for the whitelist root.
+     */
+    function leverage(
+        Order calldata order,
+        uint256 initialDeposit,
+        uint256 resultingAdditionalCollateral,
+        uint256 amountToBorrow,
+        bytes32[] memory proof
+    )
+        external
+        onlyWhitelistedBorrowers(proof)
+    {
+        // transfer initial collateral to this contract
+        uint256 collateralToPurchase = resultingAdditionalCollateral - initialDeposit;
+        COLLATERAL.safeTransferFrom(msg.sender, address(this), initialDeposit);
+
```

Comment:
Would be good to add a comment either here or in the devdocs that this contract does not support 0 leverage. In our other leverage contracts, there would be an early exit if `collateralToPurchase == 0`.